### DISCUSSION TITLE
Graphics handling

### DISCUSSION THREAD

**Discussion by Bumblebee001** [GrandOrgue/grandorgue#690] [General]
When one looks at jOrgan and GrandOrgue from a sample-set creator point of view, one may easily come to the conclusion that jOrgan is great at the way it handles graphics (creation, editing, placement on the GUI display, how image overlaps work to create "movement" eg drawing stops in and out, etc) whilst GrandOrgue is easier and better at the audio aspect.

If one can combine the two good elements basically replacing the graphics handling in GO with that of jO, GO would be so much easier and better even for end-users (graphics can be adjusted from the GUI without needing to manipulate the coding in the odf). I have long wanted to see this implemented but was sadly frowned upon by the original developers.

What I am concerned about is backward compatibility for sample-sets created for the way GO is currently coded to handle graphics.

Regards,
Mark

(PS I posted this subject via email but I don't see it in GitHub. I apologize if entering it here as a topic for discussion has duplicated the entry in some way I don't appreciate)

**Comment by oleg68** [comment]
I up this topic

**Comment by vpoguru** [comment]
The implementation of a modern way to implement beautiful screens (with overlapping controls) would be a huge step in the right direction!

**Comment by panthen** [comment]
I confess I have no experience with jOrgan, but improving the 'standard' graphics would greatly increase the appeal of GO; for instance, Piotr Grabowski's samplesets are of high audio quality, but also look fantastic.

Making things simpler and more accessible for sampleset creators and end users can only be a good thing; if it is too complicated or a hassle to make odfs (and make them look good) people won't bother and will pay for alternatives.

I understand the need for some backwards compatibility, but unless all current odfs will be broken beyond  repair then why not move forward?

**Reply by Bumblebee001** [reply]
Graphics in jOrgan do not need to have masks and be of the exact size (in case of paired images eg Off and On state of a switch). They are way simpler to create and behave better than in GO. For instance, they can easily be sized and moved around on the screen by end-user without having to go into the odf to manipulate position in terms of X and Y axis. One can literally select a group of images and move en bloc to any new / precise location (by nudging up/down/left/right) as desired. No problem! The entire interface can be configured, sized and spaced out to suit one's specific needs and saved. No fuss.

With GO it is a PITA (excuse the language). Any change has to be done in the odf. Creating images by sampleset creators is a laborious task in itself and users can definitely do little to adapt the layout on their specific screen/s. With jOrgan it is a breeze. 

I cannot be more specific as it has been a while since I did this stuff and I would need to go back and do some revision before I can explain details and I currently don't have much time at my disposal. But I think I have stated more than enough to support my request for this change to be introduced into GrandOrgue. 

As I stated: I am more concerned about compatibility between the new GO and samplesets created for the older versions. This I anticipate would be where the headaches will start!

**Reply by panthen** [reply]
Thank you for the additional detail. That sounds like quite a significant improvement for sampleset creators as well as end users.
 
When I have created odf's (although I've not made many!), I used the organ builder to organise all the locations for me and then formatted the rest myself in notepad++ instead of having to think about placing anything - so this change would be appreciated.

If such an improvement is possible, would it not be possible to create new odf's for the most popular samplesets and then if an old sampleset used, could GO not put up a message saying that it is no longer supported, and would require a previous version to load (plus link to download)? 

Also, I know this is off-topic, but I would also like to see higher resolution default textures so that even basic odf's look good.

**Reply by Bumblebee001** [reply]
You can check out my samplesets obtainable through my website: mhpo.epizy.com. When I send you the download link/s, check out the graphics. To load jOrgan sets you would of course need to install the software first, obviously, unless you have already done so. All my sets come in jO, GO and HW versions. The first set also for Aria player and Kontakt player.

**Reply by sandstone42** [reply]
For me, just eliminating the need to have the images in the same size (for example, drawknobs) would be a great update. I'm not that familiar with wxWidgets to know how to redraw just the area around the drawknob.

**Comment by larspalo** [comment]
My thoughts for the future is to reduce/minimize and hopefully eliminate any need to manually edit odfs at all. We have just slightly begun touching on the subject of having the organ settings for an organ separate from the basic odf, it would not be impossible to treat the graphical representation of an organ in a similar way to the organ settings. That is, allowing the user to override bundled settings or graphical representations of an organ, either directly from within GO or likely a better solution: a separate program, but included in the GO distribution and using the same library code. From a technical perspective it's the [Panel999] parts that easily could be broken out as the graphical rendition.

In the end we could have a much simpler sampleset creation procedure of a basic odf only stating the constructional elements of the organ, a separate settings configuration, and a separate graphical configuration. All of which could be overridden by the user if that's desired. And all done either from within GO or a different software.

However, that said, I think that people thinking about abandoning the roots of GrandOrgue and it's HW1 backwards compatibility should perhaps consider using the rights of GNU GPL and fork out into projects of their own instead of trying to defile this project. In my opinion, the GO foundation is something we can build upon, develop and adapt, but it's not called GrandOrgue anymore if we abandon the roots.

**Reply by Bumblebee001** [reply]
One of my compatriots who sadly died very recently, was the professor of international fame for "Lateral Thinking" - Prof. Edward Debono. He revolutionised the way we ought to think and his methods have been used with resounding success in all aspects of life. If we continue to think _vertically_, we become tied down or bogged down and progress will be very limited. 

Prof Debono's proven methods work side-ways - _laterally_ - hence _Lateral Thinking_ which would allow us to develop ideas or projects into far better tools than what we started with even if the ideas that are presented during brain storming sessions may appear impossible or downright ludicrous. We ought to re-think GO. Not that it is "bad" in any sense... on the contrary - it is my preferred software and the one I go for first when creating my sample-sets. This is helped, I must confess, by the existence of an easy-to-use sample-set creator, in spite of it having been left half baked by its author, Vincent Forman.

The way GO handles samples does make it a lot easier than jOrgan's need for soundfonts and a separate sound generating module like fluidsynth that has limitations; as for the graphics, it is another story. 

My vision is to create some kind of hybrid software that combines the best elements of both. There is no need to re-brand. It simply needs re-thinking in some aspects. Graphics is definitely one of them in my books. One can go for a situation where we have the current compilation and adding additional features to enhance functionality and allows sample-set creators alternative, easier options that would work in the "new revised editions" whilst allowing backward compatibility since the basics are essentially retained.

I have no knowledge of programming but I do believe that with a bit of effort and good will nothing is insurmountable in this digital age. We are certainly living through one of the most exciting times in human history. Let's continue making progress without tying ourselves down with nostalgic feelings.

I hope that my comments are considered in the same good spirit as I intended.

**Comment by Bumblebee001** [comment]
Henry Ford is quoted to have said "If I had asked people what they wanted, they would have said faster horses." We all know what he did that revolutionize transportation!

Mark
